src/desktop/space/mod.rs

```diff
@@ -0,0 +1,706 @@
+    /// To facilitate this you need to provide age values of the buffers bound to
+    /// the given `renderer`. If you stop using `Space` temporarily for rendering
+    /// or apply additional rendering operations, you need to reset the age values
+    /// accordingly as `Space` will be unable to track your custom rendering operations
+    /// to avoid rendering artifacts.
+    ///
+    /// To add aditional elements without breaking damage-tracking implement the `RenderElement`
+    /// trait and use `custom_elements` to provide them to this function. `custom_elements are rendered
+    /// after every other element.
+    ///
+    /// Returns a list of updated regions (or `None` if that list would be empty) in case of success.
+    pub fn render_output<R>(
+        &mut self,
+        renderer: &mut R,
+        output: &Output,
+        age: usize,
+        clear_color: [f32; 4],
+        custom_elements: &[DynamicRenderElements<R>],
+    ) -> Result<Option<Vec<Rectangle<i32, Logical>>>, RenderError<R>>
+    where
+        R: Renderer + ImportAll + 'static,
+        R::TextureId: 'static,
+        R::Error: 'static,
+        R::Frame: 'static,
+    {
+        if !self.outputs.contains(output) {
+            return Err(RenderError::UnmappedOutput);
+        }
+
+        type SpaceElem<R> =
+            dyn SpaceElement<R, <R as Renderer>::Frame, <R as Renderer>::Error, <R as Renderer>::TextureId>;
+
+        let mut state = output_state(self.id, output);
+        let output_size = output
+            .current_mode()
+            .ok_or(RenderError::OutputNoMode)?
+            .size
+            .to_f64()
+            .to_logical(state.render_scale)
+            .to_i32_round();
+        let output_geo = Rectangle::from_loc_and_size(state.location, output_size);
+        let layer_map = layer_map_for_output(output);
+
+        // This will hold all the damage we need for this rendering step
+        let mut damage = Vec::<Rectangle<i32, Logical>>::new();
+        // First add damage for windows gone
+        for old_toplevel in state
+            .last_state
+            .iter()
+            .filter_map(|(id, geo)| {
+                if !self
+                    .windows
+                    .iter()
+                    .map(|w| w as &SpaceElem<R>)
+                    .chain(layer_map.layers().map(|l| l as &SpaceElem<R>))
+                    .chain(custom_elements.iter().map(|c| c as &SpaceElem<R>))
+                    .any(|e| ToplevelId::from(e) == *id)
+                {
+                    Some(*geo)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<Rectangle<i32, Logical>>>()
+        {
+            slog::trace!(self.logger, "Removing toplevel at: {:?}", old_toplevel);
+            damage.push(old_toplevel);
+        }
+
+        // lets iterate front to back and figure out, what new windows or unmoved windows we have
```

Comment:
Do I understand correctly that this damage tracking does not take into account Z-order, for example a changed window will cause damage to the output even if it is completely occulted by an other, undamaged window?

If so, maybe worth letting a TODO (and maybe a tracking issue?) to not forget that this is something that has room for improvement in the future?

Reply:
Yes correct, because windows can be transparent. We need to expose `opaque_region` similar to surface damage and then iterate twice through the elements.

Once front to back to figure out, what we need to render. Then back to front to render the filtered surfaces.

But that will be another PR 😅